docs/src/examples/landau.jl

```diff
@@ -0,0 +1,265 @@
+struct ThreadCache{T, DIM, CC <: CellCache, GC <: GradientConfig, HC <: HessianConfig}
+    element_indices  ::Vector{Int}
+    element_dofs     ::Vector{T}
+    element_gradient ::Vector{T}
+    element_hessian  ::Matrix{T}
+    cellcache        ::CC
+    gradconf         ::GC
+    hessconf         ::HC
+    element_coords   ::Vector{Vec{DIM, T}}
+end
+function ThreadCache(dpc::Int, nodespercell,  args...)
+    element_indices  = zeros(Int, dpc)
+    element_dofs     = zeros(dpc)
+    element_gradient = zeros(dpc)
+    element_hessian  = zeros(dpc, dpc)
+    cellcache        = CellCache(args...)
+    gradconf         = GradientConfig(nothing, zeros(dpc), Chunk{12}())
+    hessconf         = HessianConfig(nothing, zeros(dpc), Chunk{12}())
+    coords           = zeros(Vec{3}, nodespercell)
+    return ThreadCache(element_indices, element_dofs, element_gradient, element_hessian, cellcache, gradconf, hessconf, coords )
+end
+
+# ## The Model
+# everything is combined into a model.
+mutable struct LandauModel{T, DH <: DofHandler, CH <: ConstraintHandler, TC <: ThreadCache}
+    dofs          ::Vector{T}
+    dofhandler    ::DH
+    boundaryconds ::CH
+    threadindices ::Vector{Vector{Int}}
+    threadcaches  ::Vector{TC}
+end
+
+function LandauModel(α, G, gridsize, left::Vec{DIM, T}, right::Vec{DIM, T}, elpotential) where {DIM, T}
+    grid = generate_grid(Tetrahedron, gridsize, left, right)
+    questionmark, threadindices = JuAFEM.create_coloring(grid)
+
+    qr  = QuadratureRule{DIM, RefTetrahedron}(2)
+    cvP = CellVectorValues(qr, Lagrange{DIM, RefTetrahedron, 1}())
+
+    dofhandler = DofHandler(grid)
+    push!(dofhandler, :P, 3)
+    close!(dofhandler)
+
+    dofvector = zeros(ndofs(dofhandler))
+    startingconditions!(dofvector, dofhandler)
+    boundaryconds = ConstraintHandler(dofhandler)
+    #boundary conditions can be added but aren't necessary for optimization
+    #add!(boundaryconds, Dirichlet(:P, getfaceset(grid, "left"), (x, t) -> [0.0,0.0,0.53], [1,2,3]))
+    #add!(boundaryconds, Dirichlet(:P, getfaceset(grid, "right"), (x, t) -> [0.0,0.0,-0.53], [1,2,3]))
+    close!(boundaryconds)
+    update!(boundaryconds, 0.0)
+
+    apply!(dofvector, boundaryconds)
+
+    hessian = create_sparsity_pattern(dofhandler)
+    dpc = ndofs_per_cell(dofhandler)
+    cpc = length(grid.cells[1].nodes)
+    caches = [ThreadCache(dpc, cpc, copy(cvP), ModelParams(α, G), elpotential) for t=1:nthreads()]
+    return LandauModel(dofvector, dofhandler, boundaryconds, threadindices, caches)
+end
+
+# utility to quickly save a model
+function JuAFEM.vtk_save(path, model, dofs=model.dofs)
+    vtkfile = vtk_grid(path, model.dofhandler)
+    vtk_point_data(vtkfile, model.dofhandler, dofs)
+    vtk_save(vtkfile)
+end
+
+# ## Assembly
+# This macro defines most of the assembly step, since the structure is the same for
+# the energy, gradient and Hessian calculations.
+macro assemble!(innerbody)
```

Comment:
Out of curiosity, why couldn't this just be a function?

Reply:
Oh hmm no reason. I didn't think about just passing in an innerbody "function" to the assembly function. I'll do an experiment there shouldn't be a difference, maybe it makes the example a bit more accessible?

Reply:
But do you even have to pass in a function, can't you just call this function from within the assembly function?

Reply:
Yeah, you are right, you probably need to pass a function into `assemble!` to do the "inner" part. We can leave as is, I just have a thing for trying to not use macros for as long as possible heh.

Reply:
Oh yea, I'll clean it up a bit later, I also noticed that this `CellCache` is actually not necessary using the function as a field of the `ThreadCache` also works. Before I had it since I thought for performance I could not store functions as fields, but had to do something like `(x::CellCache)(...)` as the function call for the gradient and hessian.